### ISSUE TITLE
XLLoadOptions discussion

### ISSUE THREAD

**Issue by igitur** [ClosedXML/ClosedXML#1407]
An issue to brainstorm the possible parameters in the new proposed `XLLoadOptions` (name not finalised).

- `CultureInfo` which defaults to `CultureInfo.CurrentCulture`
- `SkipPivotTables` to disable loading / saving pivot tables.
- `ThrowErrorsOnLoad`, defaults to `true`. If `false`, try to swallow errors on load
...

**Comment by Pankraty** [comment]
Instead of `ThrowErrorsOnLoad`, I'd suggest accepting `OnError` delegate, in which the user could decide whether to throw the exception, output it somewhere, or ignore. The default implementation will throw.

Besides, the external references resolver I mentioned [here](https://github.com/ClosedXML/ClosedXML/issues/1395#issuecomment-615110923) can also be passed via `XLLoadOptions` as a delegate or by some interface. (Maybe not in the next version, but some time in future).

**Comment by igitur** [comment]
Good point. Reminds me of CsvHelper's approach. We could use some of their other ideas too maybe.

**Comment by igitur** [comment]
I'm toying around with the idea of combining the current `LoadOptions` and `SaveOptions` class into a class that implements 2 interfaces:

```c#
public class XLConfiguration : IXLLoadConfiguration, IXLSaveConfiguration
```
A property like `EvaluateFormulas` could be shared between the 2 interfaces. What do you think, @Pankraty ?

**Comment by Pankraty** [comment]
So, if I understand correctly, the user will be able to do this:
```
var cfg = new XLConfiguration{ Culture = CultureInfo.GetCulture("fr-FR"))};
var wb = new XLWorkbook(path, cfg);
wb.SaveAs(otherPath); // Forgot to pass cfg - saving is done using en-US
```
In that case, the user creates a configuration that is `IXLLoadConfiguration`, `IXLSaveConfiguration` simultaneously and they may presume it will be used for both loading and saving. But since it's been passed as `IXLLoadConfiguration` it will not affect saving. Or will it? This may be a little confusing.

**Comment by igitur** [comment]
No, my idea was to store the `XLConfiguration` in the `XLWorkbook`. If the user omits it when calling `SaveAs`, then that configuration is used. If he provides it during `SaveAs`, then that provided configuration is used. If it's confusing, it might not be worth the little benefit that there is.

**Comment by Pankraty** [comment]
> No, my idea was to store the XLConfiguration in the XLWorkbook. If the user omits it when calling SaveAs, then that configuration is used. If he provides it during SaveAs, then that provided configuration is used. If it's confusing, it might not be worth the little benefit that there is.

But to be able to store the configuration during the loading we have to accept `XLConfiguration`, not `IXLLoadConfiguration`. Else, the user can pass their own implementation or `IXLLoadingConfiguration` that has nothing to do with `IXLSaveConfiguration`.

**Comment by igitur** [comment]
Yeah, true. Although, if the user implements any of our interfaces with his own class, many things will break ;-) But I think let's abandon this idea.

**Comment by igitur** [comment]
Add base date convention. See LibreOffice:
![image](https://user-images.githubusercontent.com/145854/82096544-7e6c2700-9701-11ea-85bd-3a06ae2574a0.png) and conversation at #1432

**Comment by timka3115** [comment]
Add please ability to load only needed data. In my case I needed to load excel files with a lot of cells (200 000 rows and even more) just for getting data values. I don't need work with styles, graphics and so on but internally ClosedXML do a lot of work especially for parsing styles. Like it was implemented in the Aspose Cells for .NET library, enum LoadDataFilterOptions.

**Comment by jahav** [comment]
Closing.

My vision for load options is:
* workbook culture
* object that holds hooks for volatile or external content functions (e.g. a hook for now that would be used by `DATE()` or `NOW()` or some day in the future `IMAGE()`)
* predefined number formats - they differ across cultures and there is no database of predefined formats

There won't be any selective loading - if a dev need to load only a part of a workbook or streaming, this is not a good library for that.

There won't be a switch to swallow errors or some recovery. That is something Microsoft has resources for. The plan is to try to load Excels found by Common Crawl and fix the worst offenders. That along with alignment of data structures to structure of workbook should improve a situation. If standard load can't handle it, that's it. The only "recovery" is a situation where an attribute is optional and is not parseable (e.g. some number with a value "xyz") - in that case reader will treat attribute as missing (see `XmlTreeReader.StrictAttributeParsing`).
